### PR TITLE
[spaceship] improve multiple sessions and multiple auth types

### PR DESCRIPTION
**Pull request by joshdholtz**
### Motivation and Context
`Spaceship::ConnectAPI` relied on logging in through `Spaceship::Tunes.login` or `Spaceship::Portal.login` (for web session) and or by setting `Spaceship::ConnectAPI.token = ` for JWT auth. This felt a bit janky going forward for a `Spaceship` that will slowly deprecating all `Spaceship::Tunes` and `Spaceship::Portal`.

`Spaceship::ConnectAPI` also only allowed a single client session unlike other spaceship modules.

#### Goal
- Be able to create individual `Spaceship::Connect` clients for different sessions
- Be backwards compatible with previous single session API
  - Example: `Spaceship::ConnectAPI.get_apps`
- Log in without user needing to call `Spaceship::Tunes` or `Spaceship::Portal`
- Make authoring with JWT auth easier

### Description
- `Spaceship::ConnectAPI::Client`
  - This is an aggregate of all the different `APIClient` instances 👇 
- `Spaceship::ConnectAPI::APIClient`
  - The class that handles JWT auth or subclassed to handle proxy web session auth
  - Instances
    - `Spaceship::ConnectAPI::TestFlight::Client`
    - `Spaceship::ConnectAPI::Provisioning::Client`
    - `Spaceship::ConnectAPI::Tunes::Client`
    - `Spaceship::ConnectAPI::Users::Client`
- Introduced new API module for the 4 module types:
  - New modules that get extended on client instances
    - `Spaceship::ConnectAPI::TestFlight::API`
      - Requires an API client be set on `test_flight_request_client`
    - `Spaceship::ConnectAPI::Provisioning::API`
      - Requires an API client be set on `provisioning_request_client`
    - `Spaceship::ConnectAPI::Tunes::API`
      - Requires an API client be set on `tunes_request_client`
    - `Spaceship::ConnectAPI::Users::API`
      - Requires an API client be set on `users_request_client`
  - An instance of an `APIClient` needs to get set onto each module
    - This is needed because the module defines the request path, query params, and body but does not the hostname (base URL) for web sessions
    - This allows the modules to be used by different API implementations in different levels of the clients
    - `Provisioning::API` needs a client initialized with either:
      - A `Spaceship::Portal::Client`
      - A JWT token
    - `TestFlight::API`, `Tunes::API`, `Users::API` needs a client initialized with either:
      - A `Spaceship::Tunes::Client`
      - A JWT token

### Examples

#### Singleton/Global JWT session
- Login with the global `Spaceship::ConnectAPI.auth
  - Token loaded from environment variables if not specified
- Make API calls
  - Example: `Spaceship::ConnectAPI.get_apps`
- The helpers on models use the global session
  - Example: `Spaceship::ConnectAPI::App.all`

```rb
lane :global_jwt do                                                                                                       
  fastlane_require 'spaceship'                                                                                                   
                                                                                                                                 
  if UI.confirm("Load from ENV?")                                                                                                
    # SPACESHIP_CONNECT_API_KEY_ID=ewafwefew                                                                                     
    # SPACESHIP_CONNECT_API_ISSUER_ID=wfaweewf-awef-awef-awef-awefa                                                              
    # SPACESHIP_CONNECT_API_KEY_FILEPATH=../AuthKey_FWEF32432.p8                                                                 
    Dotenv.load ".env.me_token"                                                                                                  
                                                                                                                                 
    Spaceship::ConnectAPI.auth                                                                                                   
  else                                                                                                                           
    Spaceship::ConnectAPI::Client.auth(
      key_id: "123456789",
      issuer_id: "32jflekjiowe-ewfwef-wefwef-wefwef",
      filepath: "../AuthKey_123456789.p8"
    )
  end                                                                                                                                                                                                                          

  # Uses Spaceship::ConnectAPI::TestFlight::API module
  apps = Spaceship::ConnectAPI::App.all.map(&:bundle_id).join(", ")
  puts "\tapps: #{apps}"
                                                                                                                                 
  # Uses Spaceship::ConnectAPI::Tunes::API module                                                                                               
  terrs = Spaceship::ConnectAPI::Territory.all.map(&:id).join(", ")                                                              
  puts "\tterrs: #{terrs}"                                                                                                       
                            
  # Uses Spaceship::ConnectAPI::Provisioning::API module                                                                                         
  bunds = Spaceship::ConnectAPI::BundleId.all.map(&:identifier).join(", ")                                                       
  puts "\tbunds: #{bunds}"                                                                                                       
                                  
  # Uses Spaceship::ConnectAPI::Provisioning::API module
  # Hits appstoreconnect.apple.com and uses Spaceship::Tunes style session                                                                                                                                                                                              
  users = Spaceship::ConnectAPI::User.all.map(&:username).join(", ")                                                             
  puts "\tusers: #{users}"                                                                                                       
end
```

#### Singleton/Global Web session
- This is how `Spaceship::ConnectAPI` calls were previously done
- Login with the global `Spaceship::ConnectAPI.login`
- Select team with global `Spaceship::ConnectAPI.select_team`
- Make API calls
  - Example: `Spaceship::ConnectAPI.get_apps`
- The helpers on models use the global session
  - Example: `Spaceship::ConnectAPI::App.all`

```rb

lane :asc_test_lazy_web do                                                                                                       
  fastlane_require 'spaceship'                                                                                                   
                                                                                                                                 
  if UI.confirm("Load from ENV?")        
    # FASTLANE_USER=your@email.com
    # FASTLANE_TEAM_NAME=Your Team
    Dotenv.load ".env.me_web"                                                                                                    

    Spaceship::ConnectAPI.login                                                                                                  
  elsif UI.confirm("Select team after login?")                                                                                   
    Spaceship::ConnectAPI.login("me@joshholtz.com")                                                                              
    Spaceship::ConnectAPI.select_team(team_name: "Josh Holtz")                                                                   
  else                                                                                                                           
    Spaceship::ConnectAPI.login("me@joshholtz.com", team_name: "Josh Holtz")                                                     
  end                                                                                                                            
                                                                                                                                 
  puts "=== asc_test_lazy_web"                                                                                                   
                                                                                                                                 
  # Hits appstoreconnect.apple.com and uses Spaceship::Tunes style session
  apps = Spaceship::ConnectAPI::App.all.map(&:bundle_id).join(", ")                                                              
  puts "\tapps: #{apps}"                                                                                                         

  # Hits appstoreconnect.apple.com and uses Spaceship::Tunes style session                                                                                                                                 
  terrs = Spaceship::ConnectAPI::Territory.all.map(&:id).join(", ")                                                              
  puts "\tterrs: #{terrs}"                                                                                                       

  # Hits appstoreconnect.apple.com and uses Spaceship::Portal style session                                                                                                                                 
  bunds = Spaceship::ConnectAPI::BundleId.all.map(&:identifier).join(", ")                                                       
  puts "\tbunds: #{bunds}"                                                                                                       

  # Hits appstoreconnect.apple.com and uses Spaceship::Tunes style session                                                                                                                                 
  users = Spaceship::ConnectAPI::User.all.map(&:username).join(", ")                                                             
  puts "\tusers: #{users}"                                                                                                       
end
```

#### Multiple JWT session clients
- Login with the global `Spaceship::ConnectAPI::Client.auth(key_id: “”, issuer_id: “”, filepath: “”)`
  - Token loaded from environment variables if not specified
- Make API calls on client
  - Example: `client.get_apps`
- Helpers on models will **not** use this session

```rb
lane :multiple_jwt do
  fastlane_require 'spaceship'

  awesome_client = Spaceship::ConnectAPI::Client.auth(
    key_id: "123456789",
    issuer_id: "32jflekjiowe-ewfwef-wefwef-wefwef",
    filepath: "../AuthKey_123456789.p8"
  )

  dope_client = Spaceship::ConnectAPI::Client.auth(
    key_id: "987654321",
    issuer_id: "32jflekjiowe-ewfwef-wefwef-wefwef",
    filepath: "../AuthKey_987654321.p8"
  )

  apps = awesome_client.get_apps.to_models.map(&:bundle_id).join(", ")
  puts "\t awesome apps: #{apps}"

  apps = dope_client.get_apps.to_models.map(&:bundle_id).join(", ")
  puts "\t dope apps: #{apps}"
end
```

#### Multiple web session clients
- Login with the global `Spaceship::ConnectAPI.login(user, password, team_id: “”, team_name: “”)`
  - Token loaded from environment variables if not specified
- Make API calls on client
  - Example: `client.get_apps`
- Helpers on models will **not** use this session

```rb
lane :multiple_web do
  fastlane_require 'spaceship'

  awesome_client = Spaceship::ConnectAPI::Client.login("awesome@email.com", team_name: "Team Awesome")
  dope_client = Spaceship::ConnectAPI::Client.login("dope@email.com", team_name: "Team Dope")

  apps = awesome_client.get_apps.to_models.map(&:bundle_id).join(", ")
  puts "\t awesome apps: #{apps}"

  apps = dope_client.get_apps.to_models.map(&:bundle_id).join(", ")
  puts "\t dope apps: #{apps}"
end
````

### Testing Steps

Update `Gemfile` to 👇 and run `bundle install`, `bundle update fastlane`, or `bundle update`

```rb
gem "fastlane", :git => "https://github.com/fastlane/fastlane.git", :branch => "joshdholtz-spaceship-improve-multiple-session-architecture"
```

### CODE REVIEW

**Comment by fastlane-bot-helper** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Big PR</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Review by milch — APPROVED** [review]
LGTM, just had some minor suggestions/questions

**Review by minuscorp — APPROVED** [review]
This is just awesome 🚀

File: spaceship/lib/spaceship/connect_api/api_client.rb [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+
+      protected
+
+      def with_asc_retry(tries = 5, &_block)
+        tries = 1 if Object.const_defined?("SpecHelper")
+        response = yield
+
+        status = response.status if response
+
+        if [500, 504].include?(status)
```
[the remark this answers is not in the document]
Comment (joshdholtz) [reply spaceship/lib/spaceship/connect_api/api_client.rb:158]: So this is a really weird situation 😇 It has been observed that the ASC proxy servers (the web session ones) return 500 and 504 when there are issues with the proxy server talking to the official ASC API. Some weird things happen in the middle which return errors and retrying the request works totally fine 🤷‍♂️

[the remark this answers is not in the document]
Comment (minuscorp) [reply spaceship/lib/spaceship/connect_api/api_client.rb:158]: That's a weird response from the server but I trust you👍

**Pull request merged**